src/ILLink.RoslynAnalyzer/DataFlow/DynamicallyAccessedMembersBinder.cs

```diff
@@ -0,0 +1,435 @@
+					}
+
+					yield return @event;
+				}
+
+				if ((bindingFlags & BindingFlags.DeclaredOnly) == BindingFlags.DeclaredOnly)
+					yield break;
+
+				type = type.BaseType;
+				onBaseType = true;
+			}
+		}
+
+		// declaredOnly will cause this to retrieve interfaces recursively required by the type, but doesn't necessarily
+		// include interfaces required by any base types.
+		public static IEnumerable<ITypeSymbol> GetAllInterfaceImplementations (this ITypeSymbol thisType, OperationAnalysisContext context, bool declaredOnly)
+		{
+			ITypeSymbol? type = thisType;
+			while (type != null) {
+				foreach (var i in type.Interfaces) {
+					yield return i;
+
+					ITypeSymbol? interfaceType = i;
+					if (interfaceType != null) {
+						// declaredOnly here doesn't matter since interfaces don't have base types
+						foreach (var innerInterface in interfaceType.GetAllInterfaceImplementations (context, declaredOnly: true))
+							yield return innerInterface;
+					}
+				}
+
+				if (declaredOnly)
+					yield break;
+
+				type = type.BaseType;
+			}
+		}
+
+		// Issue https://github.com/dotnet/roslyn-analyzers/issues/4845
+#pragma warning disable RS1024
+		// declaredOnly will cause this to retrieve only members of the type, not of its base types. This includes interfaces recursively
+		// required by this type (but not members of these interfaces, or interfaces required only by base types).
+		public static void GetAllOnType (this ITypeSymbol type, OperationAnalysisContext context, bool declaredOnly, List<ISymbol> members) => GetAllOnType (type, context, declaredOnly, members, new HashSet<ITypeSymbol> ());
+#pragma warning restore RS1024
+
+		static void GetAllOnType (ITypeSymbol type, OperationAnalysisContext context, bool declaredOnly, List<ISymbol> members, HashSet<ITypeSymbol> types)
+		{
+			if (!types.Add (type))
+				return;
+
+			// @TODO
+			//if (type.HasNestedTypes) {
+			//	foreach (var nested in type.NestedTypes) {
+			//		members.Add (nested);
+			//		// Base types and interfaces of nested types are always included.
+			//		GetAllOnType (nested, context, declaredOnly: false, members, types);
+			//	}
+			//}
+
+			if (!declaredOnly) {
+				var baseType = type.BaseType;
+				if (baseType != null)
+					GetAllOnType (baseType, context, declaredOnly: false, members, types);
+			}
+
+			if (!type.Interfaces.IsEmpty) {
+				if (declaredOnly) {
+					foreach (var iface in type.GetAllInterfaceImplementations (context, declaredOnly: true))
+						members.Add (iface);
+				} else {
+					foreach (var iface in type.Interfaces) {
+						members.Add (iface);
+						var interfaceType = iface;
+						if (interfaceType == null)
+							continue;
+						GetAllOnType (interfaceType, context, declaredOnly: false, members, types);
+					}
+				}
+			}
+
+			foreach (var member in type.GetMembers ()) {
+				if (member is IFieldSymbol f)
```

Comment:
I would replace the ifs with a switch. I think you can also use member.Kind which might (?) be a bit faster than the type check, here and elsewhere.

Reply:
I replaced this with a switch and changed the if patterns as suggested by the IDE code fixes. But I didn't change all the if statements primarily to make it easier to see the 1:1 port with the linker.